File: Minimize the Heights II.cpp
class Solution {
  public:
    int getMinDiff(vector<int> &arr, int k) {
        sort(arr.begin(),arr.end());
        int n=arr.size();
        int ans=arr[n-1]-arr[0];
        int l=arr[0]+k;
        int r=arr[n-1]-k;
        for(int i=0;i<(n-1);i++){
            int mini=min(l,arr[i+1]-k);
            int maxi=max(r,arr[i]+k);
            if(mini>=0)
            ans=min(ans,maxi-mini);
        }
        return ans;
    }
};

File: Minimize the Heights II.java
// gfg 160 problem today(23/11/24) 

class Solution {
    int getMinDiff(int[] arr, int k) {
        // code here
        int n = arr.length;
        
        Arrays.sort(arr);

        int min = arr[n-1] - arr[0];

        for (int i = 1; i < n; i++) {
            if (arr[i] - k < 0) {
                continue;
            }

            int currentMin = Math.max(arr[n-1] - k, arr[i-1] + k) - Math.min(arr[0] + k, arr[i] - k);
            min = Math.min(min, currentMin);
        }

        return min;
    }
}
